### PR TITLE
usability improvements

### PR DESCRIPTION
**Pull request by gdkrmr**
- Runs the server in background
- restarts the server at every `pprof()` call.
- `@pprof ex` as an alias to `Profile.clear(); @profile ex; pprof()`

### CODE REVIEW

File: src/PProf.jl [hunk 1]
```diff
@@ -217,8 +217,35 @@ function pprof(data::Union{Nothing, Vector{UInt}} = nothing,
...
+            kill(PProf.pprof_proc[1])
+        end
+
+        PProf.pprof_proc[1] = run(`$go_pprof -http=$webhost:$webport $out`, wait = false)
+    end
+
     out
 end
 
+const pprof_proc = Union{Base.Process, Nothing}[nothing]
```
Comment (vchuravy) [inline src/PProf.jl]: this should be a `Ref`

File: src/PProf.jl [hunk 2]
```diff
@@ -217,8 +217,35 @@ function pprof(data::Union{Nothing, Vector{UInt}} = nothing,
...
+
+Alias for
+```julia
+@profile expr
+pprof()
+```
+"""
+macro pprof(ex)
+    quote
+        Profile.clear()
```
Comment (vchuravy) [inline src/PProf.jl]: I would re-export clear from `PProf` and drop it from this macro. It often can be useful to accumulate more data and so the `clear` should be conscious.

File: src/PProf.jl [hunk 3]
```diff
@@ -217,8 +217,35 @@ function pprof(data::Union{Nothing, Vector{UInt}} = nothing,
...
     end
 
-    if web run(`$go_pprof -http=$webhost:$webport $out`) end
+    if web
+        if !isnothing(PProf.pprof_proc[1])
+            kill(PProf.pprof_proc[1])
+        end
+
+        PProf.pprof_proc[1] = run(`$go_pprof -http=$webhost:$webport $out`, wait = false)
+    end
```
Comment (NHDaly) [inline src/PProf.jl]: I think it would be nice to move this into a function, something like "refresh_pprof()", so that if the user updates the output file manually, they can refresh the server without having to rerun `pprof()`, which would overwrite the file.

File: src/PProf.jl [hunk 4]
```diff
@@ -217,8 +217,35 @@ function pprof(data::Union{Nothing, Vector{UInt}} = nothing,
...
+pprof()
+```
+"""
+macro pprof(ex)
+    quote
+        Profile.clear()
+        @profile $(esc(ex))
+        pprof()
+    end
+end
```
[the remark this answers is not in the document]
Comment (vchuravy) [reply src/PProf.jl]: Macros are weird man. If you do the global escape you definitely need the interpolation. If you only do the escape locally I don't think you will need it, but better to check.

[the remark this answers is not in the document]
Comment (gdkrmr) [reply src/PProf.jl]: I took this from `@time`. Let's see...

File: src/PProf.jl (hunk 2 above)
Comment (gdkrmr) [reply src/PProf.jl]: clear is not exported by `Profile` so I will just import it, this way you can just call `PProf.clear()`, the same way you would call `Profile.clear()`

**Comment by gdkrmr** [comment]
- imported `Profile.clear`, so it can be called as `PProf.clear`
- `PProf.refresh(...)` restarts the server
- `pprof_proc` -> `PProf.proc`
- Documentation.

File: src/PProf.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (gdkrmr) [reply src/PProf.jl]: The local escape works for me without `using Profile`. Not sure what else could go wrong, Should I do the global escape anyways?

File: test/PProf.jl [hunk 5]
```diff
@@ -78,4 +79,18 @@ end
...
+    @test PProf.proc[] !== nothing
+
+    PProf.kill()
+    @test PProf.proc[] === nothing
+
+    @test isfile("profile.pb.gz")
+    rm("profile.pb.gz")
+end
+
 end
```
Comment (NHDaly) [inline test/PProf.jl]: @gdkrmr  Can you please also add a couple tests for the new `refresh()` and `kill()` behavior? Maybe just verify that the old process is in fact killed after each function is called?

Something like:
```julia
current_proc = PProf.proc[]
@assert isopen(current_proc)
PProf.refresh()
@test !isopen(current_proc)
@test isopen(PProf.proc[])

current_proc = PProf.proc[]
@assert isopen(current_proc)
PProf.kill()
@test !isopen(current_proc)
@test !isopen(PProf.proc[])
```

File: src/PProf.jl [hunk 6]
```diff
@@ -217,8 +228,56 @@ function pprof(data::Union{Nothing, Vector{UInt}} = nothing,
...
+"""
+function refresh(; webhost::AbstractString = "localhost",
+                   webport::Integer = 57599,
+                   file::AbstractString = "profile.pb.gz")
+
+    if proc[] !== nothing
+        Base.kill(proc[])
+    end
+
+    proc[] = run(`$go_pprof -http=$webhost:$webport $file`, wait = false)
```
[the remark this answers is not in the document]
Comment (NHDaly) [reply src/PProf.jl]: I've asked about this on Discourse:
https://discourse.julialang.org/t/how-to-run-a-process-in-background-but-still-close-it-on-exit/27231

Hopefully we can get some more definitive explanation on this.

File: src/PProf.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (gdkrmr) [reply src/PProf.jl]: done

File: src/PProf.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (gdkrmr) [reply src/PProf.jl]: changed it, if the discussion on discourse results in a better way, reopen.

File: test/PProf.jl (hunk 5 above)
Comment (gdkrmr) [reply test/PProf.jl]: `isopen` does not really work for processes, I used `process_running` and `process_exited`.

**Comment by gdkrmr** [comment]
- the process is started by `open(pipe(...))` now. this prints an extra error message in the terminal
`"Main binary filename not available."`
but works otherwise
- added unit tests for managing the subprocess.

File: test/PProf.jl [hunk 7]
```diff
@@ -78,4 +79,55 @@ end
...
+
+    @test isfile("profile.pb.gz")
+    rm("profile.pb.gz")
 end
+
+@testset "subprocess refresh" begin
+
+    let arr = []
+        @pprof foo(10000, 5, arr)
+        sleep(2)
```
Comment (NHDaly) [inline test/PProf.jl]: why do we need to `sleep()` in this test (here and below)? `@pprof` shouldn't return until the process is started, right?

File: test/PProf.jl [hunk 8]
```diff
@@ -78,4 +79,55 @@ end
...
+    end
+
+    current_proc = PProf.proc[]
+    @test process_running(current_proc)
+
+    PProf.kill()
+    sleep(2)
```
Comment (NHDaly) [inline test/PProf.jl]: Same here, this shouldn't return until the process has exited, right?

**Comment by NHDaly** [comment]
> this prints an extra error message in the terminal
"Main binary filename not available."
but works otherwise

I think this is because `pprof` wants you to also pass the name of the binary that the profile belongs to, so that it can use that for extra debug output or something. It gives the same warning if you just run `pprof` on the command line:
```bash
$ deps/usr/bin/pprof -http=localhost:23432 profile.pb.gz
Main binary filename not available.
Serving web UI on http://localhost:23432
```

One noticeable change is that now when you run `pprof()` it doesn't automatically open the server in the browser.. you have to copy/paste the link. I think this seems less than ideal... I can't yet figure out how to fix that.


Also, annoyingly, somehow the `pprof` process is _still_ staying alive! It's not us; i've confirmed that the `pprof` process is running in our process group (which means it should be killed when we exit), but somehow it's jumping ship and staying alive, moving its parent process to `1`. I'm not very familiar with unix land, but this sounds like some weird unix process stuff...

It's confusing because if i start a simple child process like this, it _does_ get closed when I exit:
```julia
p = open(pipeline(`bash -c 'while true; do sleep 1 && echo "hi"; done'`))  # killed when we ctrl-D julia
```
but the go subprocess is not:
```julia
p = open(pipeline(`$(PProf.go_pprof) -http=localhost:23432 profile.pb.gz`))  # _NOT_ killed when we ctrl-D julia!
```
Some googling led me to this article, but i don't think this actually has anything to do with Go (unless the binary is somehow doing something to set itself to "please don't kill me!!" or something):
https://medium.com/@felixge/killing-a-child-process-and-all-of-its-children-in-go-54079af94773

> - added unit tests for managing the subprocess.
> - squashed all commits.

🙌  Thanks! :) They look great

So to summarize my remaining complaints:
- [ ] `pprof()` now doesn't open the webserver in the browser. I think it would be nice to do this (though if you already have it open, you maybe _don't_ want to do that... Ideally we'd only do this the first time?
- [ ] The `pprof` binary is still outliving julia, which means you can't run PProf again without manually finding and killing that process (because it complains about `bind: address already in use`).

The second one seems problematic enough that we probably shouldn't merge this until it's fixed. :'( I'm sorry, @gdkrmr! Thanks for your really fast and great work on this; i think this will make the package more usable. :)

**Comment by NHDaly** [comment]
> * [ ]  The `pprof` binary is still outliving julia, which means you can't run PProf again without manually finding and killing that process (because it complains about `bind: address already in use`).

Oh, huh, apparently my understanding of unix process death was just totally wrong. Orphaned processes are reparented by default, not killed:
https://stackoverflow.com/questions/284325/how-to-make-child-process-die-after-parent-exits

The more you know!

And what's more, it doesn't sound like there's a good way to kill them by default, which is super lame. The best solutions people have are to have your child process poll the parent process and exit when the parent exits... But this sounds nontrivial.

Maybe a better approach will be to:
  1.  register an atexit hook to kill the child process, and
  2. Allow pprof() to retry different random ports if the default port is occupied?

**Comment by gdkrmr** [comment]
>pprof() now doesn't open the webserver in the browser. I think it would be nice to do this (though if you already have it open, you maybe don't want to do that... Ideally we'd only do this the first time?

I think the best behaviour is to not open the browser.
- Implementing this in a platform and browser independent way sounds tedious.
- Most terminals detect links anyway and you can just click on it.

**Comment by NHDaly** [comment]
For the second problem, i just wrote this kind of dumb utility, which maybe would be sufficient for our usecase?
https://github.com/NHDaly/ExitWithParent.jl

**Comment by gdkrmr** [comment]
> For the second problem, i just wrote this kind of dumb utility, which maybe would be sufficient for our usecase?
> https://github.com/NHDaly/ExitWithParent.jl

That is kind of cool :-), maybe a bit overkill.

File: test/PProf.jl (hunk 7 above)
Comment (gdkrmr) [reply test/PProf.jl]: I just copied the thing from above in the file.

Comment (gdkrmr) [reply test/PProf.jl]: the sleep after `refresh()` and `kill()` is actually necessary for the tests to run reliably.

**Comment by NHDaly** [comment]
Okay! I've addressed my concerns by just adding an `atexit` hook, and that seems to work pretty well! :)

If for some reason it doesn't work (like if julia crashes) the user can always just provide a different `webport`, so i think that's good enough! :)

Thanks for the brainstorming with me, @gdkrmr! And for this nice PR :)

**Comment by NHDaly** [comment]
Oh, we should probably update the README as well, with the new usage

**Comment by NHDaly** [comment]
Alright, thanks @gdkrmr! :) I'm merging this now! :)
